design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+#### The way it is now
+
+Right now each of the atlas types has a reference within the render loop (discussed below).  They also seem to do their image uploading on the rendering thread.  That's unnecessary even in OpenGL, but it is simple.
+
+#### Required Changes
+
+The rendering loop shouldn't need to know about any of the atlases. That bookkeeping is better done as a callback.  That would open up the possibility of doing more atlas types in the future.
+
+Further, the atlases should try to do their data uploading on other threads.  If you're doing this with OpenGL you need to open shared contexts and admittedly, it can get a bit complex.  With Metal it's much simpler.
+
+Much of this could be done with expanded observers as discussed below.
+
+#### Benefits
+
+Less work done in the rendering loop is always better.  More CPU available for UI developers to update the curves on their buttons.
+
+Making the atlas updates more generic might allow for new types, or even the fabled Dynamic Texture Atlas.  We have one in WhirlyGlobe-Maply and it's great for adding random geometry and textures into the mix.  But that's outside the scope of this proposal.
+
+### Main Render Loops
+
+_Addresses refactoring that make migration easier, particularly our [north stars](#north-stars)._
+
+Every real time rendering toolkit has a main rendering loop or something like it.  It's not actually a loop anymore, it's a callback.  Years ago there actually was a main rendering loop where the program in question would draw a frame, wait for a defined period and then draw another frame.  With modern OS' and the rise of modern UI toolkits, that loop has been replaced by time or event based callbacks.  We will still sometimes refer to it as a "rendering loop", however.
+
+On iOS this runs on the main thread and on Android this runs on its own separate thread.  You might think the latter is better, but it has its own problems (gesture lag).  In any case, MapLibre Native handles both just fine.
+
+The toolkit's main rendering "loop" lives in [renderer_impl](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/renderer_impl.cpp) and it's worth a look.  It serves as a central point to begin examining the real time rendering.
+
+As you might suspect, the rendering loop logic and implementation are heavily influenced by OpenGL.
+
+#### The way it is now
+
+Since to toolkit supports one rendering SDK, OpenGL ES, there is only one rendering "loop" and it focuses on having Layers render themselves.
+
+#### Required Changes
+
+The most obvious change is we'll need one renderer_impl per rendering SDK.  We can follow the backend/renderer logic to implement that for OpenGL with just a bit of tweaking and then open it up for Metal and Vulkan.
```

Comment:
Is that really required? Do you mean we will need to copy&paste the renderer_impl for each SDK?

Maybe we can build an abstraction here.

Reply:
Also we probably only need one renderer_impl for OpenGL and one for low-overhead-apis (Vulkan, Metal), because Vulkan and Metals interface are more or less identical.

Reply:
Interesting, do you have a drawing for this? I am a bit lost here...

Reply:
By the way the pngs you already have look beautiful. How did you make them?

Reply:
I'm drawn back to this because it's the key issue.  

What we found in our toolkit (WhirlyGlobe-Maply) is that the rendering "loop" and specific Drawable implementation gets very, very specific indeed.  As our Metal implementation got more advanced, we moved from direct drawing to indirect drawing and from allocating individual memory buffers to using Heaps.

Each of those changes required extensive work in the Metal Drawables and Metal Rendering "loop", but nowhere else.  Only when we added a new fundamental concept, did both Metal and OpenGL interfaces have to change.

Reply:
Damon made the figures from my whiteboard scrawls.  They look great!

Reply:
> I'm drawn back to this because it's the key issue.

Sorry I don't understand this sentence. Do you mean this question is too specific?

Reply:
No, it meant that I came back to the question later to try to explain a bit more.  It was an interesting question.